kernel/kernel-api/src/test/scala/io/delta/kernel/TransactionSuite.scala

```diff
@@ -75,6 +78,53 @@ class TransactionSuite extends AnyFunSuite with VectorTestUtils with MockEngineU
       })
     }
   }
+
+  Seq(true, false).foreach { icebergCompatV2Enabled =>
+    test(s"generateAppendActions: iceberg comaptibily checks, " +
+      s"icebergCompatV2Enabled=$icebergCompatV2Enabled") {
+      val txnState = testTxnState(testSchema, enableIcebergCompatV2 = icebergCompatV2Enabled)
+      val engine = testMockEngine(testSchema)
+
+      Seq(
+        // missing stats
+        (
+          testDataFileStatuses(
+            "file1" -> testStats(Some(10)), // valid stats
+            "file2" -> None // missing stats
+          ),
+          "Iceberg V2 compatibility requires statistics" // expected error message
+        )
+      ).foreach { case (actionRows, expectedErrorMsg) =>
+        if (icebergCompatV2Enabled) {
+          val ex = intercept[KernelException] {
+            generateAppendActions(engine, txnState, actionRows, testDataWriteContext())
+              .forEachRemaining(_ => ()) // consume the iterator
+          }
+          assert(ex.getMessage.contains(expectedErrorMsg))
+        } else {
+          // when icebergCompatV2Enabled is disabled, no exception should be thrown
+          generateAppendActions(engine, txnState, actionRows, testDataWriteContext())
+            .forEachRemaining(_ => ()) // consume the iterator
+        }
+      }
+
+      // valid stats
+      val dataFileStatuses = testDataFileStatuses(
+        "file1" -> testStats(Some(10)),
+        "file2" -> testStats(Some(20))
+      )
+      var actStats: Seq[String] = Seq.empty
+      generateAppendActions(engine, txnState, dataFileStatuses, testDataWriteContext())
+        .forEachRemaining { addActionRow =>
+          val addOrdinal = addActionRow.getSchema.indexOf("add")
+          val add = addActionRow.getStruct(addOrdinal)
+          val statsOrdinal = add.getSchema.indexOf("stats")
+          actStats = actStats :+ add.getString(statsOrdinal)
+        }
+
+      assert(actStats === Seq("{\"numRecords\":10}", "{\"numRecords\":20}"))
```

Comment:
Do we have any E2E tests where we do have the delta statistics present (i.e. where we are verifying that Spark can read these stats)?

Reply:
Currently not explicitly. The current e2e tests will generate the stats and are read by Spark for verification. As part of the last PR (once all these iceberg compat v2 checks are done), will explicitly check.